src/bio/smd/tests/smd_ut.c

```diff
@@ -244,9 +251,95 @@ ut_pool(void **state)
 	assert_int_equal(rc, -DER_NONEXIST);
 }
 
+static void
+ut_dev_replace(void **state)
+{
+	struct smd_dev_info	*dev_info, *tmp_dev;
+	struct smd_pool_info	*pool_info, *tmp_pool;
+	uuid_t			 dev_id3, pool_id1, pool_id2;
+	d_list_t		 pool_list, dev_list;
+	uint64_t		 blob_id;
+	int			 i, rc, pool_cnt = 0, dev_cnt = 0;
+
+	uuid_generate(dev_id3);
+	uuid_generate(pool_id1);
+	uuid_generate(pool_id2);
+
+	/* Assign pools, they were unassigned in prior pool test */
+	for (i = 0; i < 4; i++) {
+		rc = smd_pool_assign(pool_id1, i, i << 10, 100);
+		assert_int_equal(rc, 0);
+
+		rc = smd_pool_assign(pool_id2, i, i << 20, 200);
+		assert_int_equal(rc, 0);
+	}
+
+	D_INIT_LIST_HEAD(&pool_list);
+	rc = smd_pool_list(&pool_list, &pool_cnt);
+	assert_int_equal(rc, 0);
+	assert_int_equal(pool_cnt, 2);
+
+	/* Assign different blobs to dev1's targets 0, 1, 2 */
+	d_list_for_each_entry(pool_info, &pool_list, spi_link) {
+		pool_info->spi_blobs[0] = 555;
+		pool_info->spi_blobs[1] = 666;
+		pool_info->spi_blobs[2] = 777;
+	}
+
+	/* Replace dev1 with dev3 without marking dev1 as faulty */
+	rc = smd_dev_replace(dev_id1, dev_id3, &pool_list);
+	assert_int_equal(rc, -DER_INVAL);
+
+	rc = smd_dev_set_state(dev_id1, SMD_DEV_FAULTY);
+	assert_int_equal(rc, 0);
+
+	/* Replace dev1 with dev2 */
+	rc = smd_dev_replace(dev_id1, dev_id2, &pool_list);
+	assert_int_equal(rc, -DER_INVAL);
+
+	/* Replace dev1 with dev3 */
+	rc = smd_dev_replace(dev_id1, dev_id3, &pool_list);
+	assert_int_equal(rc, 0);
+
+	/* Verify device after replace */
+	D_INIT_LIST_HEAD(&dev_list);
+	rc = smd_dev_list(&dev_list, &dev_cnt);
+	assert_int_equal(rc, 0);
+	assert_int_equal(dev_cnt, 2);
+
+	d_list_for_each_entry_safe(dev_info, tmp_dev, &dev_list, sdi_link) {
```

Comment:
For lists where you are deleting the list, it's a bit more natural to use d_list_pop_entry in a while loop (e.g. while ((dev_info = d_list_entry_pop(...)) != NULL) { )